File: CDC/1.py
## WAP to implement Lexical Analyzer to identify tokens.

import re

# Define token patterns
token_specification = [
    ('NUMBER',    r'\d+(\.\d*)?'),  # Integer or decimal number
    ('ASSIGN',    r'='),            # Assignment operator
    ('END',       r';'),            # Statement terminator
    ('ID',        r'[A-Za-z]+'),    # Identifiers
    ('OP',        r'[+\-*/]'),      # Arithmetic operators
    ('SKIP',      r'[ \t]'),        # Skip over spaces and tabs
    ('MISMATCH',  r'.'),            # Any other character
]

# Compile the regex for performance
token_re = '|'.join(f'(?P<{pair[0]}>{pair[1]})' for pair in token_specification)
get_token = re.compile(token_re).match

# Token class to store token information
class Token:
    def __init__(self, type, value, position):
        self.type = type
        self.value = value
        self.position = position

    def __repr__(self):
        return f'Token({self.type}, {self.value}, {self.position})'

def lex(text):
    line_no = 1
    line_start = 0
    pos = 0
    tokens = []
    match = get_token(text)

    while match is not None:
        type = match.lastgroup
        value = match.group(type)
        if type == 'NUMBER':
            value = float(value) if '.' in value else int(value)
        elif type == 'ID' and value in {'if', 'else', 'for', 'while'}:
            type = value.upper()
        elif type == 'SKIP':
            pos = match.end()
            match = get_token(text, pos)
            continue
        elif type == 'MISMATCH':
            raise RuntimeError(f'{value!r} unexpected on line {line_no}')
        tokens.append(Token(type, value, match.start()))
        pos = match.end()
        match = get_token(text, pos)

    tokens.append(Token('EOF', '', pos))
    return tokens

# Example usage
text = input('Enter the lexeme: ')
tokens = lex(text)
for token in tokens:
    print(token)

File: CDC/1.c
// WAP to implement Lexical Analyzer to identify tokens.

#include <stdio.h>
#include <ctype.h>
#include <string.h>

// Function to check if a character is a valid identifier character
int isIdentifierChar(char ch) {
    return isalnum(ch) || ch == '_';
}

// Function to check if a character is a valid operator character
int isOperatorChar(char ch) {
    return strchr("+-*/=", ch) != NULL;
}

// Function to tokenize the input string
void tokenize(char* input) {
    int len = strlen(input);
    int i = 0;

    while (i < len) {
        // Skip whitespace characters
        if (isspace(input[i])) {
            i++;
            continue;
        }

        // Check if the current character is an identifier
        if (isalpha(input[i]) || input[i] == '_') {
            int j = i + 1;
            while (j < len && isIdentifierChar(input[j])) {
                j++;
            }
            printf("Identifier: %.*s\n", j - i, &input[i]);
            i = j;
            continue;
        }

        // Check if the current character is an operator
        if (isOperatorChar(input[i])) {
            printf("Operator: %c\n", input[i]);
            i++;
            continue;
        }

        // Check if the current character is a number
        if (isdigit(input[i])) {
            int j = i + 1;
            while (j < len && isdigit(input[j])) {
                j++;
            }
            printf("Number: %.*s\n", j - i, &input[i]);
            i = j;
            continue;
        }

        // If none of the above, it's an invalid character
        printf("Invalid character: %c\n", input[i]);
        i++;
    }
}

int main() {
    char input[100];

    printf("Enter input string: ");
    fgets(input, sizeof(input), stdin);

    tokenize(input);

    return 0;
}